cmd/snap/cmd_connections.go

```diff
@@ -0,0 +1,195 @@
+)
+
+type cmdConnections struct {
+	clientMixin
+	All         bool `long:"all"`
+	Positionals struct {
+		Snap installedSnapName
+	} `positional-args:"true"`
+}
+
+var shortConnectionsHelp = i18n.G("List interface connections")
+var longConnectionsHelp = i18n.G(`
+The connections command lists connections between plugs and slots
+in the system.
+
+Unless <snap> is provided, the listing is for connected plugs and
+slots for all snaps in the system. In this mode, pass --all to also
+list unconnected plugs and slots.
+
+$ snap connections <snap>
+
+Lists connected and unconnected plugs and slots for the specified
+snap.
+`)
+
+func init() {
+	addCommand("connections", shortConnectionsHelp, longConnectionsHelp, func() flags.Commander {
+		return &cmdConnections{}
+	}, map[string]string{
+		"all": i18n.G("Show connected and unconnected plugs and slots"),
+	}, []argDesc{{
+		// TRANSLATORS: This needs to be wrapped in <>s.
+		name: "<snap>",
+		// TRANSLATORS: This should not start with a lowercase letter.
+		desc: i18n.G("Constrain listing to a specific snap"),
+	}})
+}
+
+func isSystemSnap(snap string) bool {
+	return snap == "core" || snap == "snapd" || snap == "system"
+}
+
+func endpoint(snap, name string) string {
+	if isSystemSnap(snap) {
+		return ":" + name
+	}
+	return snap + ":" + name
+}
+
+type connection struct {
+	slot          string
+	plug          string
+	interfaceName string
+	manual        bool
+	gadget        bool
+}
+
+func (cn connection) String() string {
+	opts := []string{}
+	if cn.manual {
+		opts = append(opts, "manual")
+	}
+	if cn.gadget {
+		opts = append(opts, "gadget")
+	}
+	if len(opts) == 0 {
+		return "-"
+	}
+	return strings.Join(opts, ",")
+}
+
+type byConnectionData []connection
+
+func (b byConnectionData) Len() int      { return len(b) }
+func (b byConnectionData) Swap(i, j int) { b[i], b[j] = b[j], b[i] }
+func (b byConnectionData) Less(i, j int) bool {
+	iCon, jCon := b[i], b[j]
+	if iCon.interfaceName != jCon.interfaceName {
+		return iCon.interfaceName < jCon.interfaceName
+	}
+	if iCon.plug != jCon.plug {
+		return iCon.plug < jCon.plug
+	}
+	return iCon.slot < jCon.slot
+}
+
+func (x *cmdConnections) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+
+	opts := client.ConnectionOptions{
+		All: x.All,
+	}
+	wanted := string(x.Positionals.Snap)
+	if wanted != "" {
+		if x.All {
+			// passing a snap name already implies --all, error out
+			// when it was passed explicitly
+			return fmt.Errorf(i18n.G("cannot use --all with snap name"))
+		}
+		// when asking for a single snap, include its disconnected plugs
+		// and slots
+		opts.Snap = wanted
+		opts.All = true
+		// print all slots
+		x.All = true
+	}
+
+	connections, err := x.client.Connections(&opts)
+	if err != nil {
+		return err
+	}
+	if len(connections.Plugs) == 0 && len(connections.Slots) == 0 {
+		return nil
+	}
+
+	notes := make([]connection, 0, len(connections.Established)+len(connections.Undesired))
```

Comment:
we might want a different name here now,  s/notes/conns/ ?   s/notes/annotatedConns/ ?